File: app/models/document.py
import uuid
from sqlalchemy import Column, DateTime, String, Integer, Float, Boolean, func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
from app.db.session import Base


class Document(Base):
    """SQLAlchemy model representing the documents metadata table.

    Serves as the foundation for parsed, partitioned, or indexed files.
    """

    __tablename__ = "documents"

    id = Column(
        UUID(as_uuid=True),
        primary_key=True,
        default=uuid.uuid4,
        index=True,
        nullable=False,
    )
    original_filename = Column(String, nullable=False)
    stored_filename = Column(String, nullable=False)
    bucket_name = Column(String, nullable=False)
    storage_path = Column(String, nullable=False)
    mime_type = Column(String, nullable=False)
    file_size = Column(Integer, nullable=False)
    upload_status = Column(String, default="pending", nullable=False)
    description = Column(String, nullable=True)
    
    # Document Processing Pipeline Fields
    processing_status = Column(String, default="Uploaded", nullable=False)
    processing_started_at = Column(DateTime(timezone=True), nullable=True)
    processing_completed_at = Column(DateTime(timezone=True), nullable=True)
    processing_duration = Column(Float, nullable=True)
    processing_error = Column(String, nullable=True)
    parser_used = Column(String, nullable=True)
    retry_count = Column(Integer, default=0, nullable=False)
    page_count = Column(Integer, nullable=True)
    character_count = Column(Integer, nullable=True)
    word_count = Column(Integer, nullable=True)
    line_count = Column(Integer, nullable=True)
    extracted_text_version = Column(Integer, default=1, nullable=False)
    extraction_completed = Column(Boolean, default=False, nullable=False)

    extracted_text = relationship(
        "ExtractedText", uselist=False, back_populates="document", cascade="all, delete-orphan"
    )

    created_at = Column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    updated_at = Column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )

File: frontend/src/types/document.ts
export interface Document {
  id: string;
  original_filename: string;
  stored_filename: string;
  bucket_name: string;
  storage_path: string;
  mime_type: string;
  file_size: number;
  upload_status: string;
  description: string | null;
  
  // Document Processing Pipeline Fields
  processing_status: string;
  processing_started_at: string | null;
  processing_completed_at: string | null;
  processing_duration: number | null;
  processing_error: string | null;
  parser_used: string | null;
  retry_count: number;
  page_count: number | null;
  character_count: number | null;
  word_count: number | null;
  line_count: number | null;
  extracted_text_version: number;
  extraction_completed: boolean;

  created_at: string;
  updated_at: string;

  // Document Semantic Indexing Pipeline Fields
  chunk_count: number | null;
  embedding_model: string | null;
  embedding_dimension: number | null;
  embedding_status: string | null;
  index_status: string | null;
  indexed_at: string | null;
  vector_collection: string | null;
  indexing_duration: number | null;
  failed_chunk_count: number;
}

export interface PaginatedDocumentResponse {
  items: Document[];
  total: number;
  skip: number;
  limit: number;
}

export interface UploadDocumentRequest {
  file: File;
  title?: string;
  description?: string;
}

export interface DocumentUploadStatus {
  id: string;
  file: File;
  progress: number;
  status: 'pending' | 'uploading' | 'completed' | 'error';
  error?: string;
}
